### PR TITLE
Add vehicles from Mapping to simulation when using SumoScenarioAmbassador

### PR DESCRIPTION
**Pull request by schwepmo**
## Type of this PR 

- [ ] Bug fix
- [x] Enhancement

## Description

* This PR adds functionality to simultaneous add vehicles via SUMOs route files and MOSAICs mapping files
* Additionally a very basic Integration Test has been added to showcase and test this functionality
* MOSAIC vehicle types will now be written in an additional-file instead of and route-file (doesn't change functionality)

## Issue(s) related to this PR

 * Resolves internal issue #86

## Affected parts of the online documentation

* The SUMO documentation will be adapted to prop

## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/util/SumoRouteFileCreator.java [hunk 1]
```diff
@@ -59,7 +59,7 @@
 
     private final static Logger log = LoggerFactory.getLogger(SumoRouteFileCreator.class);
 
-    final static String VEHICLE_TYPE_ROUTE_FILE_NAME = "mosaic_types.rou.xml";
+    final static String MOSAIC_TYPES_FILE_NAME = "mosaic_types.add.xml";
```
Comment (kschrab) [inline fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/util/SumoRouteFileCreator.java:62]: Now you could rename the class 😄 , e.g. `SumoVehicleTypesWriter`

**Review by kschrab — CHANGES_REQUESTED** [review]
Overall very clean solution! Some minor changes requested.

File: test/scenarios/MosaicAndSumoVehicles/sumo/tiergarten.rou.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  ~ http://www.eclipse.org/legal/epl-2.0
+  ~
+  ~ SPDX-License-Identifier: EPL-2.0
+  ~
+  ~ Contact: mosaic@fokus.fraunhofer.de
+  -->
+<routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://sumo.dlr.de/xsd/routes_file.xsd">
+    <vType id="SumoVehicle"/>
+    <vehicle id="0" type="SumoVehicle" depart="10" >
+        <route edges="4068038_251150126_428788319_21487167 4068038_251150126_428788319_272365223 4068038_428788319_408194194_428788319 4068038_428788319_408194194_272256206 4068038_408194194_423839224_408194194 4068038_408194194_423839224_26738489 4068038_408194194_423839224_313006383 4068038_423839224_26704448_423839224 36337928_26704448_27537750_26704448 4609244_27537750_27537749_27537750 4609243_27537749_252864801_27537749 4609243_252864801_252864802_252864801 4609243_252864801_252864802_265786533 32935479_252864802_21487170_252864802 32935479_252864802_21487170_341364279 32935479_252864802_21487170_248919692 32935479_252864802_21487170_82654385 32935479_252864802_21487170_27011308 32935479_252864802_21487170_341364277 32935479_252864802_21487170_341364270 32935479_252864802_21487170_27011305" />
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply test/scenarios/MosaicAndSumoVehicles/sumo/tiergarten.rou.xml:18]: Idk if there is a way to get route IDs from SUMO-Gui, but I called `getRouteIds()` via TraCI and interestingly enough the IDs of the routes get the  use the vehicle-ID prepended with an `!`. So in the case of vehicle `0`, its routes' ID is `!0`.

[the remark this answers is not in the document]
Comment (kschrab) [reply test/scenarios/MosaicAndSumoVehicles/sumo/tiergarten.rou.xml:18]: Thats very comfortable :)

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoScenarioAmbassador.java [hunk 3]
```diff
@@ -69,7 +73,17 @@
...
+
+    /**
+     * Set containing all vehicles, that have been added using the SUMO route file.
+     */
+    private final Set<String> vehiclesAddedViaRouteFile = new HashSet<>();
+
+    /**
+     * Set containing all vehicles, that have been added using the Mapping file.
+     */
+    private final Set<String> vehiclesAddedViaMapping = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoScenarioAmbassador.java:86]: good point!

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/util/SumoRouteFileCreator.java (hunk 1 above)
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/util/SumoRouteFileCreator.java:62]: I like.

File: bundle/src/assembly/resources/scenarios/LuST/README.md [hunk 4]
```diff
@@ -6,6 +6,6 @@ This MOSAIC-scenario integrates the Luxembourg SUMO Traffic (LuST) Scenario to M
 * Adjust the `mapping_config.json` to get your apps mapped onto the available vehicle types
 * Execute MOSAIC with provided `runtime.xml`  :
```
Comment (kschrab) [inline bundle/src/assembly/resources/scenarios/LuST/README.md:7]: this line should be removed too

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java [hunk 5]
```diff
@@ -64,15 +68,30 @@
...
+     */
+    private final Set<String> vehiclesAddedViaRouteFile = new HashSet<>();
+
+    /**
+     * Set containing all vehicles, that have been added from the RTI e.g. using the Mapping file.
+     */
+    private final Set<String> vehiclesAddedViaRti = new HashSet<>();
+
+    /**
+     * Instance of {@link SumoVehicleTypesWriter} used to write routes to a *.rou.xml file.
```
Comment (kschrab) [inline fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java:92]: this comment is now wrong

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java [hunk 6]
```diff
@@ -362,28 +428,38 @@ private boolean isTruckOrTrailer(VehicleClass vehicleClass) {
...
         String subDir = new File(sumoConfig.sumoConfigurationFile).getParent();
         if (StringUtils.isNotBlank(subDir)) {
             dir = new File(dir, subDir);
         }
         // keep single instance
-        if (sumoRouteFileCreator == null) {
-            sumoRouteFileCreator = new SumoRouteFileCreator(dir, sumoConfig);
+        if (sumoVehicleTypesWriter == null) {
+            sumoVehicleTypesWriter = new SumoVehicleTypesWriter(dir, sumoConfig);
         }
```
Comment (kschrab) [inline fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java:463]: do we still need this extra method/field? we only use the SumoVehicleTypesWriter once now, aren't we? we could move this to `writeTypesFromMapping` (and rename the method to `writeTypesFromRti`)

File: lib/mosaic-network/src/main/java/org/eclipse/mosaic/lib/coupling/NetworkEntityIdTransformer.java [hunk 7]
```diff
@@ -86,4 +81,10 @@ public String fromExternalId(Integer externalId) throws IllegalStateException {
         }
         return nodeId;
     }
+
+    @Override
+    public void reset() {
+        idMap = HashBiMap.create();
+        nextId = new AtomicInteger();
```
Comment (kschrab) [inline lib/mosaic-network/src/main/java/org/eclipse/mosaic/lib/coupling/NetworkEntityIdTransformer.java:88]: you could still keep the fields final by using `clear()` and `.set(0)`

**Review by kschrab — CHANGES_REQUESTED** [review]
Still some minor changes requested. You can remove the "draft" label.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java (hunk 6 above)
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java:463]: true, I missed this

**Review by kschrab — COMMENTED** [review]
Code looks fine now, however, the new integration test fails on ci.eclipse.org/mosaic

**Comment by kschrab** [comment]
> Code looks fine now, however, the new integration test fails on ci.eclipse.org/mosaic

```
2021-02-02 08:26:04,983 ERROR MappingAmbassador - Configuration object could not be instantiated: 
java.lang.InstantiationException: The CMappingAmbassador config is not valid: [19,22][/vehicles/0/route] The value must be of type in [string], but actual type is integer.

	at org.eclipse.mosaic.lib.util.objects.ObjectInstantiation.validateFile(ObjectInstantiation.java:241)
	at org.eclipse.mosaic.lib.util.objects.ObjectInstantiation.read(ObjectInstantiation.java:146)
	at org.eclipse.mosaic.lib.util.objects.ObjectInstantiation.readFile(ObjectInstantiation.java:104)
	at org.eclipse.mosaic.lib.util.objects.ObjectInstantiation.readFile(ObjectInstantiation.java:86)
	at org.eclipse.mosaic.fed.mapping.ambassador.MappingAmbassador.<init>(MappingAmbassador.java:82)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.eclipse.mosaic.starter.MosaicSimulation.loadFederate(MosaicSimulation.java:337)
	at org.eclipse.mosaic.starter.MosaicSimulation.loadFederates(MosaicSimulation.java:303)
	at org.eclipse.mosaic.starter.MosaicSimulation.runSimulation(MosaicSimulation.java:179)
	at org.eclipse.mosaic.test.junit.MosaicSimulationRule.executeSimulation(MosaicSimulationRule.java:154)
	at org.eclipse.mosaic.test.junit.MosaicSimulationRule.executeSimulation(MosaicSimulationRule.java:130)
	at org.eclipse.mosaic.test.junit.MosaicSimulationRule.executeSimulation(MosaicSimulationRule.java:125)
	at org.eclipse.mosaic.test.junit.MosaicSimulationRule.executeTestScenario(MosaicSimulationRule.java:117)
	at org.eclipse.mosaic.test.MosaicAndSumoVehiclesIT.runSimulation(MosaicAndSumoVehiclesIT.java:41)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

**Review by kschrab — APPROVED** [review]

**Pull request merged**